File: app/src/main/java/com/unsupportedpastels/hermesandroid/relay/RelayFoldersClient.kt
package com.unsupportedpastels.hermesandroid.relay

import com.unsupportedpastels.hermesandroid.gateway.HermesChatMethodNotFoundException
import com.unsupportedpastels.hermesandroid.gateway.HermesChatProtocolException
import com.unsupportedpastels.hermesandroid.gateway.HostDirectoryEntry
import com.unsupportedpastels.hermesandroid.gateway.HostDirectoryListing
import com.unsupportedpastels.mercury.core.files.RelayFoldersContract
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject

class RelayFoldersUnsupportedException : HermesChatProtocolException(
    RelayFoldersContract.unsupportedMessage,
)

class RelayFoldersInvalidInputException : HermesChatProtocolException(
    "Choose an absolute server folder and a valid single folder name.",
)

class RelayFoldersInvalidResponseException : HermesChatProtocolException(
    "The Relay host returned an invalid folder listing.",
)

/**
 * Capability-gated Relay folder operations. The request callback must use the
 * already-admitted chat controller; this class never opens a second connection.
 */
class RelayFoldersClient {
    private val mutex = Mutex()
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun list(
        profile: String,
        path: String?,
        ensureCurrent: suspend () -> Unit = {},
        request: suspend (String, JsonObject) -> JsonObject,
    ): HostDirectoryListing = mutex.withLock {
        val params = parseParams(
            RelayFoldersContract.listParams(profile, path),
        )
        perform(
            method = RelayFoldersContract.listMethod,
            params = params,
            request = request,
            ensureCurrent = ensureCurrent,
        )
    }

    suspend fun create(
        profile: String,
        parentPath: String,
        name: String,
        ensureCurrent: suspend () -> Unit = {},
        request: suspend (String, JsonObject) -> JsonObject,
    ): HostDirectoryListing = mutex.withLock {
        val params = parseParams(
            RelayFoldersContract.createParams(profile, parentPath, name),
        )
        // This method is deliberately one-shot. The caller must not retry after
        // an ambiguous transport failure because the host may already have made
        // the directory.
        perform(
            method = RelayFoldersContract.createMethod,
            params = params,
            request = request,
            ensureCurrent = ensureCurrent,
        )
    }

    private suspend fun perform(
        method: String,
        params: JsonObject,
        request: suspend (String, JsonObject) -> JsonObject,
        ensureCurrent: suspend () -> Unit,
    ): HostDirectoryListing {
        ensureCurrent()
        val status = try {
            request("relay.status", buildJsonObject {})
        } catch (_: HermesChatMethodNotFoundException) {
            throw RelayFoldersUnsupportedException()
        }
        ensureCurrent()
        if (!RelayFoldersContract.supports(status.toString())) {
            throw RelayFoldersUnsupportedException()
        }

        ensureCurrent()
        val result = try {
            // Drain a dispatched RPC before releasing the shared reader permit;
            // cancellation is checked again after returning to the parent context.
            withContext(NonCancellable) { request(method, params) }
        } catch (_: HermesChatMethodNotFoundException) {
            throw RelayFoldersUnsupportedException()
        }
        currentCoroutineContext().ensureActive()
        ensureCurrent()
        val decoded = RelayFoldersContract.decodeListing(result.toString())
            ?: throw RelayFoldersInvalidResponseException()
        return HostDirectoryListing(
            path = decoded.path,
            directories = decoded.entries.map { HostDirectoryEntry(it.name, it.path) },
            parentPath = decoded.parentPath,
            root = decoded.root,
            lockedRoot = decoded.lockedRoot,
            canChangePath = decoded.canChangePath,
        )
    }

    private fun parseParams(value: String?): JsonObject =
        value?.let { runCatching { json.parseToJsonElement(it) as? JsonObject }.getOrNull() }
            ?: throw RelayFoldersInvalidInputException()
}

File: ios/Mercury/MercuryKit/Files/RelayFoldersClient.swift
import Foundation
import MercuryCore

/// Folder operations are Mercury-owned, capability-gated RPCs, never private
/// Hermes routes or a fallback to a saved direct-mode origin.
enum RelayFoldersError: Error, Equatable, LocalizedError {
    case unsupported
    case invalidInput
    case invalidResponse
    case hostRejected(String)

    var errorDescription: String? {
        switch self {
        case .unsupported:
            return RelayFoldersContract.shared.unsupportedMessage
        case .invalidInput:
            return "Choose an absolute server folder and a valid single folder name."
        case .invalidResponse:
            return "The Relay host returned an invalid folder listing."
        case let .hostRejected(message):
            return message
        }
    }
}

@MainActor
final class RelayFoldersClient {
    typealias Request = @MainActor (String, [String: Any]) async throws -> [String: Any]
    private let profile: String
    private let request: Request

    init(profile: String, request: @escaping Request) {
        self.profile = profile
        self.request = request
    }

    func list(path: String? = nil) async throws -> HostFileListing {
        guard let json = RelayFoldersContract.shared.listParams(profile: profile, path: path) else {
            throw RelayFoldersError.invalidInput
        }
        return try await perform("relay.folders.list", paramsJSON: json)
    }

    func createDirectory(parentPath: String, name: String) async throws -> HostFileListing {
        guard let json = RelayFoldersContract.shared.createParams(
            profile: profile, parentPath: parentPath, name: name
        ) else { throw RelayFoldersError.invalidInput }
        // Do not retry mutations automatically after an ambiguous transport error.
        return try await perform("relay.folders.create", paramsJSON: json)
    }

    private func perform(_ method: String, paramsJSON: String) async throws -> HostFileListing {
        try Task.checkCancellation()
        let status: [String: Any]
        do { status = try await request("relay.status", [:]) }
        catch is ChatMethodNotFoundError { throw RelayFoldersError.unsupported }
        guard RelayFoldersContract.shared.supports(statusJson: try json(status)) else {
            throw RelayFoldersError.unsupported
        }
        try Task.checkCancellation()
        guard let data = paramsJSON.data(using: .utf8),
              let params = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RelayFoldersError.invalidInput
        }
        let result: [String: Any]
        do { result = try await request(method, params) }
        catch is ChatMethodNotFoundError { throw RelayFoldersError.unsupported }
        try Task.checkCancellation()
        guard let listing = RelayFoldersContract.shared.decodeListing(json: try json(result)) else {
            throw RelayFoldersError.invalidResponse
        }
        return HostFileListing(
            path: listing.path,
            entries: listing.entries.map { HostFileEntry(name: $0.name, path: $0.path, isDirectory: true) },
            parentPath: listing.parentPath,
            root: listing.root,
            lockedRoot: listing.lockedRoot,
            canChangePath: listing.canChangePath
        )
    }

    private func json(_ value: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: value)
        guard let value = String(data: data, encoding: .utf8) else {
            throw RelayFoldersError.invalidResponse
        }
        return value
    }
}
